framework/aster-frame/src/mm/page/meta.rs

```diff
@@ -0,0 +1,266 @@
+#[repr(u8)]
+pub enum PageUsage {
+    // The zero variant is reserved for the unused type. Only an unused page
+    // can be designated for one of the other purposes.
+    Unused = 0,
+    /// The page is reserved or unusable. The kernel should not touch it.
+    Reserved = 1,
+
+    /// The page is used as a frame, i.e., a page of untyped memory.
+    Frame = 32,
+    /// The page is used as the head frame in a segment.
+    SegmentHead = 33,
+
+    /// The page is used by a page table.
+    PageTable = 64,
+    /// The page stores metadata of other pages.
+    Meta = 65,
+    /// The page stores the kernel such as kernel code, data, etc.
+    Kernel = 66,
+}
+
+#[repr(C)]
+pub(in crate::mm) struct MetaSlot {
+    /// The metadata of the page.
+    ///
+    /// The implementation may cast a `*const MetaSlot` to a `*const PageMeta`.
+    _inner: MetaSlotInner,
+    /// To store [`PageUsage`].
+    pub(super) usage: AtomicU8,
+    pub(super) ref_count: AtomicU32,
+}
+
+pub(super) union MetaSlotInner {
+    frame: ManuallyDrop<FrameMeta>,
+    seg_head: ManuallyDrop<SegmentHeadMeta>,
+    // Make sure the the generic parameters don't effect the memory layout.
+    pt: ManuallyDrop<PageTablePageMeta<PageTableEntry, PagingConsts>>,
+}
+
+// Currently the sizes of the `MetaSlotInner` union variants are no larger
+// than 8 bytes and aligned to 8 bytes. So the size of `MetaSlot` is 16 bytes.
+//
+// Note that the size of `MetaSlot` should be a multiple of 8 bytes to prevent
+// cross-page accesses.
+const_assert_eq!(size_of::<MetaSlot>(), 16);
+
+/// All page metadata types must implemented this sealed trait,
+/// which ensures that each fields of `PageUsage` has one and only
+/// one metadata type corresponding to the usage purpose. Any user
+/// outside this module won't be able to add more metadata types
+/// and break assumptions made by this module.
+///
+/// If a page type needs specific drop behavior, it should specify
+/// when implementing this trait. When we drop the last handle to
+/// this page, the `on_drop` method will be called.
+pub trait PageMeta: Default + Sync + private::Sealed + Sized {
+    const USAGE: PageUsage;
+
+    fn on_drop(page: &mut Page<Self>);
+}
+
+mod private {
+    pub trait Sealed {}
+}
+
+// ======= Start of all the specific metadata structures definitions ==========
+
+use private::Sealed;
+
+#[derive(Debug, Default)]
+#[repr(C)]
+pub struct FrameMeta {}
+
+impl Sealed for FrameMeta {}
+
+#[derive(Debug, Default)]
+#[repr(C)]
+pub struct SegmentHeadMeta {
+    /// Length of the segment in bytes.
+    pub(in crate::mm) seg_len: u64,
+}
+
+impl Sealed for SegmentHeadMeta {}
+
+impl From<Page<FrameMeta>> for Page<SegmentHeadMeta> {
+    fn from(page: Page<FrameMeta>) -> Self {
+        // FIXME: I intended to prevent a page simultaneously managed by a segment handle
+        // and a frame handle. However, `Vmo` holds a frame handle while block IO needs a
+        // segment handle from the same page.
```

Comment:
Are you sure that having `SegmentHeadMeta` is the most wise design decision?

As I understand it, the main motivation of introducing `SegmentHeadMeta` is to make creating a `Segment` more efficient: with `SegmentHeadMeta`, only one metadata slot needs to be updated; without it, `N` metadata slots of `FrameMeta` must be updated, where `N` is the number of frames in the segment. But I think that the benefit of this optimization is neglectable as the user code allocating a segment almost always need to initialize the pages within the segment, whose cost is already `O(N)` and much greater than the cost of updating the metadata. In short, `SegmentHeadMeta` does not seem to have a strong motivation.

Now let's consider the downside of introducing `SegmentHeadMeta`. Having `SegmentHeadMeta` is not simply a matter of adding a new type of metadata slot, but one of upgrading the whole metadata slot mechanism. This is because all other existing or foreseeable page metadata types can be handled one page/slot at a time. But `SegmentHeadMeta` must treat multiple pages/slots as whole.

For example, let's consider a `SegmentHeadMeta` positioned at the `N`-th slot, having a length of `M` frames. Since only the head slot is updated, should the subsequent slots, from `N+1` to `N+M-1`, be considered in use or not? Should a call to `Page::from_unused` for the corresponding pages succeed or not? This is undefined.

The philosophy of Asterinas is to value simplicity, correctness, and safety over performance.

Reply:
I am not sure indeed. This is a surrender over the current usage of `VmSegments`. To address the problem that `Page::from_unused` would succeed on tail pages, we can introduce `SegmentTailMeta` and initialize/destroy all the tail meta when creating or releasing the segment.